tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,300 @@
+#![cfg_attr(not(feature = "sync"), allow(dead_code, unreachable_pub))]
+
+use super::Semaphore;
+use crate::loom::cell::UnsafeCell;
+use std::cell::Cell;
+use std::fmt;
+use std::future::Future;
+use std::mem::MaybeUninit;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// A thread-safe cell which can be written to only once.
+///
+/// Provides the functionality to set the value by using an async
+/// function via [`OnceCell::get_or_init`]
+///
+/// [`OnceCell::get_or_init`]: crate::sync::OnceCell::get_or_init
+///
+/// # Examples
+/// ```
+/// use tokio::sync::OnceCell;
+/// use std::pin::Pin;
+/// use std::future::Future;
+///
+/// async fn some_computation() -> u32 {
+///     1 + 1
+/// }
+///
+/// fn call_async_fn() -> Pin<Box<dyn Future<Output = u32> + Send>> {
+///     Box::pin(some_computation())
+/// }
+///
+/// #[tokio::main]
+/// async fn main() {
+///     static ONCE: OnceCell<u32> = OnceCell::new();
+///     let result1 = tokio::spawn(async {
+///         ONCE.get_or_init(call_async_fn).await
+///     }).await.unwrap();
+///     assert_eq!(*result1, 2);
+/// }
+pub struct OnceCell<T> {
+    value_set: AtomicBool,
+    value: UnsafeCell<MaybeUninit<T>>,
+    sema: Semaphore,
```

Comment:
I think it is more readable like this.
```suggestion
    semaphore: Semaphore,
```